### ISSUE TITLE
No actual implementations of 2.4.9 "Link Purpose (Link Only)"?

### ISSUE THREAD

**Issue by Jookia** [w3c/wcag#1119]
[WCAG 2.1 2.4.9 "Link Purpose (Link Only)"](https://www.w3.org/TR/WCAG21/#link-purpose-link-only) says this:

> A mechanism is available to allow the purpose of each link to be identified from link text alone, except where the purpose of the link would be ambiguous to users in general.

[Understanding Success Criterion 2.4.9: Link Purpose (Link Only)](https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-link-only.html) elaborates:

> Best practice is that links with the same destination would have the same descriptions, but links with different purposes and destinations would have different descriptions (see also Success Criterion 3.2.4 which calls for consistency in identifying components that have the same functionality). Because the purpose of a link can be identified from its link text, links can be understood when they are out of context, such as when the user agent provides a list of all the links on a page. 

> The word "mechanism" is used to allow authors to either make all links fully understandable out of context by default or to provide a way to make them this way. This is done because for some pages, making the links all unambiguous by themselves makes the pages easier for some users and harder for others. Providing the ability to make the links unambiguous (by them selves) or not provides both users with disabilities with the ability to use the page in the format that best meets their needs.

> For example: A page listing 100 book titles along with links to download the books in HTML, PDF, DOC, TXT, MP3, or AAC might ordinarily be viewed as the title of the book as a link with the words "in HTML" after it. then the sentence "Also available in: " followed by a series of short links with text of "HTML", "PDF", "DOC", "TXT", "MP3", and "AAC". At Level 3, some users could opt to view the page this way - because they would find the page harder to understand or slower to use if the full title of the book were included in each of the links. Others could opt to view the page with the full title as part of each of the links so that each link was understandable in itself. Both the former and the latter groups could include people with visual or cognitive disabilities that used different techniques to browse or that had different types or severities of disability.

Then it lists the following techniques for implementing this:

1. [ARIA8: Using aria-label for link purpose](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA8.html)
2. [Providing a control near the beginning of the Web page that changes the link text](https://www.w3.org/WAI/WCAG21/Techniques/general/G189.html)
3. [Using CSS to hide a portion of the link text](https://www.w3.org/WAI/WCAG21/Techniques/css/C7.html)
4. [Using scripts to change the link text](https://www.w3.org/WAI/WCAG21/Techniques/client-side-script/SCR30.html)

This all seems reasonable, so I'd like to implement it on my website. But after searching online and examining various websites I have never actually seen this implemented or read about implementations let alone seen code.

Do people even implement this?

**Comment by ArnaudDelafosse** [comment]
Hi @Jookia This Technique [Using CSS to hide a portion of the link text](https://www.w3.org/WAI/WCAG21/Techniques/css/C7.html) is extremely common and widely used.

**Comment by Jookia** [comment]
Oh, do you have any examples I could see of it being used to achieve
this goal?

**Comment by ArnaudDelafosse** [comment]
What we recommend for giving context to identical/non-explicit links while doing accessibility consulting: 
`<a href="...">Read more<span class="sr-only"> on So and so</span></a>`  where "So and so" is the sting in the previous `<hn>` (the article excerpt's heading for example)
if that's not possible for some reason, this will do:
`<a href="..." title="Read more on So and so">Read more</a>`.

**Comment by Jookia** [comment]
That makes sense. But what mechanism is used to toggle showing/hiding
it? How does that work?

**Comment by patrickhlauke** [comment]
as an aside, i'll note that most sites out in the wild only go up to WCAG AA, not AAA. so they mostly aim to satisfy 2.4.4 rather than  2.4.9. And while 2.4.4's understanding mentions various users groups (motion impairment, cognitive disabilities, ...) it mostly gets checked/evaluated (at least from my experience) in the context of assistive technology users (who will get things like appropriately visually hidden extra text announced when they reach the link, without the need for any mechanism to make it actually visible).

https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context

**Comment by ArnaudDelafosse** [comment]
@Jookia if you meant how does that `sr-only` CSS class _work_, here's the lowdown: https://www.scottohara.me/blog/2017/04/14/inclusively-hidden.html#hiding-content-visually

@patrickhlauke said it all. If your _Read more_ link is visually close to its related text, then context is pretty much there for someone with regular eyesight. However for someone who cannot see the text and navigating via a list of links for example, having a screen reader vocalise a link with its context is very useful. 
There's a very good and thorough article about what a blind person has to go through to understand non explicit links but it's in French (you might wanna try G Translate) https://access42.net/lecteur-decran-liens-lire-informations-contexte

**Comment by Jookia** [comment]
No I mean how do you add a mechanism to toggle the sr-only CSS hidden or
shown for people that want to read context and people who don't?

**Comment by JAWS-test** [comment]
> No I mean how do you add a mechanism to toggle the sr-only CSS hidden or shown for people that want to read context and people who don't?

Although this is a recommended technique, I have never seen this

**Comment by ArnaudDelafosse** [comment]
Just like @JAWS-test, I don't think I've ever seen this implemented. 
However, if you go for the `title` attribute option (see my 1st answer), _some_ sighted people will see it. Kind of.

**Comment by awkawk** [comment]
You can also meet this SC by just making sure that the links make sense out of context, as described in https://www.w3.org/WAI/WCAG21/Techniques/general/G91.

**Comment by Jookia** [comment]
For a concrete example, I'm making a blog theme. Each post has a list of
tags and categories. Right now the list can look like 'Tags: example,
tags, here' with each being a link which satisfies in context links but
fails out of context links.

But having it as 'Tags: posts tagged example, posts tagged tags, posts
tagged here' creates a lot of noise for people reading it who can
understand in context links, including screen readers.

It's sounding like creating a mechanism to toggle showing or hiding the
verbose part isn't actually people do? The suggestions in the thread so
far are to make links verbose always (that doesn't work in this case) or
only verbose for screen readers (so blind people have to hear verbose
links whether they like it or not)

It's coming across to me that it doesn't really matter to always fully
comply with this guideline since there's no really any implementation of
it or really any people asking for implementations of it for all cases.

**Comment by fstrr** [comment]
This issue looks like it can be closed. If it needs to be re-opened, please do that and convert it to a Discussion.
